### DISCUSSION TITLE
why memory.usage read from file but not compute from stats.

### DISCUSSION THREAD

**Discussion by zdyj3170101136** [opencontainers/runc#4154] [General]
in containerv1, read memory stats from file memory.stats, read memory usage from file memory.usage_in_bytes.

why do not compute memory usage from memory.stats[rss] + memory.stats[cache]?

it there have any difference between value with memory.usage_in_bytes and memory.stats[rss] + memory.stats[cache]?

**Comment by cyphar** [comment]
They measure different things. `memory.usage_in_bytes` in cgroupv1 and `memory.usage` in cgroupv2 account for **all** system memory usage by processes in the cgroups, which includes shmem files, dirty pages, writeback, etc. None of these are accounted for in `rss` and `cache` (or the cgroupv2 equivalents) -- if you look at `runc events --stat` of most containers, `rss+cache` will be smaller than the usage reported by the kernel.

As for why we read them from the file, it's because we just provide the information the kernel gives you. It's up to the user as to what information they need -- if they need `rss+cache` they can calculate that themselves from the stats we provide.
